sound/soc/sof/intel/mtl.c

```diff
@@ -0,0 +1,941 @@
+		if (!hlink->ref_count)
+			snd_hdac_ext_bus_link_power_down(hlink);
+	}
+
+	/* check dma status and clean up CORB/RIRB buffers */
+	if (!bus->cmd_dma_state)
+		snd_hdac_bus_stop_cmd_io(bus);
+#endif
+
+	return ret;
+}
+
+static int mtl_dsp_resume(struct snd_sof_dev *sdev)
+{
+	const struct sof_dsp_power_state target_state = {
+		.state = SOF_DSP_PM_D0,
+		.substate = SOF_HDA_DSP_PM_D0I0,
+	};
+	int ret;
+
+	ret = mtl_resume(sdev, false);
+	if (ret < 0)
+		return ret;
+
+	return snd_sof_dsp_set_power_state(sdev, &target_state);
+}
+
+static int mtl_dsp_runtime_resume(struct snd_sof_dev *sdev)
+{
+	const struct sof_dsp_power_state target_state = {
+		.state = SOF_DSP_PM_D0,
+	};
+	int ret;
+
+	ret = mtl_resume(sdev, true);
+	if (ret < 0)
+		return ret;
+
+	return snd_sof_dsp_set_power_state(sdev, &target_state);
+}
+
+/* Meteorlake ops */
+struct snd_sof_dsp_ops sof_mtl_ops = {
+	/* probe and remove */
+	.probe		= hda_dsp_probe,
+	.remove		= hda_dsp_remove,
+	.shutdown	= hda_dsp_shutdown,
+
+	/* Register IO */
+	.write		= sof_io_write,
+	.read		= sof_io_read,
+	.write64	= sof_io_write64,
+	.read64		= sof_io_read64,
+
+	/* Block IO */
+	.block_read	= sof_block_read,
+	.block_write	= sof_block_write,
+
+	/* doorbell */
+	.irq_thread	= mtl_ipc_irq_thread,
+
+	/* ipc */
+	.send_msg	= mtl_ipc_send_msg,
+	.get_mailbox_offset = mtl_dsp_ipc_get_mailbox_offset,
+	.get_window_offset = mtl_dsp_ipc_get_window_offset,
+
+	.ipc_msg_data	= hda_ipc_msg_data,
+	.set_stream_data_offset = hda_set_stream_data_offset,
+
+	/* machine driver */
+	.machine_select = hda_machine_select,
+	.machine_register = sof_machine_register,
+	.machine_unregister = sof_machine_unregister,
+	.set_mach_params = hda_set_mach_params,
+
+	/* debug */
+	.debug_map	= mtl_dsp_debugfs,
+	.debug_map_count	= ARRAY_SIZE(mtl_dsp_debugfs),
+	.dbg_dump	= mtl_dsp_dump,
+	.ipc_dump	= cnl_ipc_dump,
+
+	/* stream callbacks */
+	.pcm_open	= hda_dsp_pcm_open,
+	.pcm_close	= hda_dsp_pcm_close,
+	.pcm_hw_params	= hda_dsp_pcm_hw_params,
+	.pcm_hw_free	= hda_dsp_stream_hw_free,
+	.pcm_trigger	= hda_dsp_pcm_trigger,
+	.pcm_pointer	= hda_dsp_pcm_pointer,
+
+	/* firmware loading */
+	.load_firmware = snd_sof_load_firmware_raw,
+
+	/* pre/post fw run */
+	.pre_fw_run = mtl_dsp_pre_fw_run,
+	.post_fw_run = mtl_dsp_post_fw_run,
+
+	/* firmware run */
+	.run = mtl_dsp_cl_boot_firmware,
+
+	/* dsp core get/put */
+	.core_get = mtl_dsp_core_power_up,
+	.core_put = mtl_dsp_core_power_down,
+
+	/* trace callback */
+	.trace_init = hda_dsp_trace_init,
+	.trace_release = hda_dsp_trace_release,
+	.trace_trigger = hda_dsp_trace_trigger,
+
+	/* DAI drivers */
+	.drv		= skl_dai,
+	.num_drv	= SOF_SKL_NUM_DAIS,
+
+	/* PM */
+	.suspend		= mtl_dsp_suspend,
+	.resume			= mtl_dsp_resume,
+	.runtime_suspend	= mtl_dsp_runtime_suspend,
+	.runtime_resume		= mtl_dsp_runtime_resume,
+	.runtime_idle		= hda_dsp_runtime_idle,
+	.set_hw_params_upon_resume = hda_dsp_set_hw_params_upon_resume,
```

Comment:
i have a vague memory this routine was no longer needed?

Reply:
> i have a vague memory this routine was no longer needed?

[.set_hw_params_upon_resume = hda_dsp_set_hw_params_upon_resume,](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda-common-ops.c#L95) is also set in `sof_hda_common_ops`. So I think it is still needed.